File: CmsAdminService/src/main/java/com/demirturk/cms/service/CategoryService.java
package com.demirturk.cms.service;

import com.demirturk.cms.exception.CmsException;
import com.demirturk.cms.model.dto.CategoryDto;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public interface CategoryService {
    /**
     *
     * @return
     * @throws CmsException
     */
    List<CategoryDto> getAll() throws CmsException;

    /**
     *
     * @param id
     * @return
     * @throws CmsException
     */
    List<CategoryDto> findFirstLevelCategoriesByCategoryTemplateId(Long id) throws CmsException;

    /**
     *
     * @param id
     * @return
     * @throws CmsException
     */
    List<CategoryDto> findAllByParentCategoryId(Long id) throws CmsException;

    /**
     *
     * @param categoryDto
     * @return
     * @throws CmsException
     */
    CategoryDto create(CategoryDto categoryDto) throws CmsException;

    /**
     *
     * @param categoryDto
     * @param id
     * @return
     * @throws CmsException
     */
    CategoryDto update(CategoryDto categoryDto, Long id) throws CmsException;

    /**
     *
     * @param id
     * @return
     * @throws CmsException
     */
    boolean delete(Long id) throws CmsException;
}

File: cmsAdminFrontend/src/app/common/service/category.service.ts
import {Injectable} from '@angular/core';
import {HttpHeaders} from '@angular/common/http';
import {ApiService} from '../../shared/services/api.service';
import {map} from 'rxjs/operators';
import {Observable} from 'rxjs';
import {BaseCrudService} from '../../shared/services/baseCrud.service';

@Injectable({
  providedIn: 'root'
})
export class CategoryService implements BaseCrudService{
  private CATEGORY_PATH = '/category';
  private FIND_ALL_BY_PARENT_CATEGORY = this.CATEGORY_PATH + '/findAllByParentCategoryId';
  private FIND_FIRST_LEVEL_CATEGORIES_BY_CATEGORY_TEMPLATE = this.CATEGORY_PATH + '/findFirstLevelCategoriesByCategoryTemplateId';

  httpOptions = {
    headers: new HttpHeaders(
      {'Content-Type': 'application/json'}
    )
  };

  constructor(private apiService: ApiService) {

  }

  getFirstCategoriesByCategoryTemplateId(id): Observable<any> {
    return this.apiService.post(this.FIND_FIRST_LEVEL_CATEGORIES_BY_CATEGORY_TEMPLATE, id).pipe(map(
      response => {
        if (response) {
          return response;
        } else {
          console.log(response);
          return null;
        }
      }
    ));
  }
  getFindAllByParentCategory(id): Observable<any> {
    return this.apiService.post(this.FIND_ALL_BY_PARENT_CATEGORY, id).pipe(map(
      response => {
        if (response) {
          return response;
        } else {
          console.log(response);
          return null;
        }
      }
    ));
  }


  checkUrl(url): Observable<any> {
    return undefined;
  }

  create(object): Observable<any> {
    return this.apiService.post(this.CATEGORY_PATH, object).pipe(map(
      response => {
        if (response) {
          return response;
        } else {
          console.log(response);
          return null;
        }
      }
    ));
  }

  delete(id): Observable<any> {
    return undefined;
  }

  findByUrl(url): Observable<any> {
    return undefined;
  }

  getAll(): Observable<any> {
    return undefined;
  }

  update(object): Observable<any> {
    return this.apiService.put(this.CATEGORY_PATH + '/' + object.id , object).pipe(map(
      response => {
        if (response) {
          return response;
        } else {
          console.log(response);
          return null;
        }
      }
    ));
  }
}
